### PR TITLE
Bump transformers version range

### PR DESCRIPTION
**Pull request by shchur**
*Issue #, if available:* #5428

*Description of changes:*
- Change transformers version range to `>=4.51,<4.58` due to breaking changes in v4.51-4.57 affecting AutoMM
- Update AutoMM code to handle transformers breaking changes:
  1. **Remove `return_dict` parameter from CLIP model calls** (`clip.py`)
     - Transformers 4.57 removed this parameter from `CLIPVisionTransformer.forward()` and `CLIPTextTransformer.forward()`, the model always returns dict by default
     - https://github.com/huggingface/transformers/pull/36794
  2. **Don't set config sub-attributes to None** (`clip.py`)
     - Removed `self.config.vision_config = None` and `self.config.text_config = None`
     - Transformers 4.57's config saving fails when sub-configs are None
  3. **Fix gradient checkpointing** (`hf_text.py`, `t_few.py`)
     - Changed `gradient_checkpointing_enable()` to `gradient_checkpointing_enable(gradient_checkpointing_kwargs={"use_reentrant": False})`
     - Required for PyTorch 2.x + transformers 4.57 to avoid "backward through graph twice" errors
     - https://github.com/huggingface/transformers/issues/26969

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by shchur** [comment]
/platform_tests 34208cd7e4470ca0f4dc4eaa8451d8f3eb17ce2a
[Platform Tests Output][1]

[1]: https://github.com/autogluon/autogluon/actions/runs/19852226382

File: core/src/autogluon/core/_setup_utils.py [hunk 1]
```diff
@@ -31,7 +31,7 @@
     "torch": ">=2.6,<2.10",  # Major version cap, sync with common/src/autogluon/common/utils/try_import.py. torchvision version in multimodelal/setup.py can effectively constrain version as well
     "lightning": ">=2.5.1,<2.6",  # Major version cap
     "async_timeout": ">=4.0,<6",  # Major version cap
-    "transformers[sentencepiece]": ">=4.38.0,<4.50",  # there is a breaking change in 4.50 for model config saving
+    "transformers[sentencepiece]": ">=4.51.0,<4.58",  # lower bound because of a breaking change in 4.51
```
Comment (abdulfatir) [inline core/src/autogluon/core/_setup_utils.py:34]: What's the breaking change?

File: multimodal/src/autogluon/multimodal/models/hf_text.py [hunk 2]
```diff
@@ -130,7 +130,7 @@ def __init__(
 
         self.gradient_checkpointing = gradient_checkpointing
         if gradient_checkpointing:
-            self.model.gradient_checkpointing_enable()
+            self.model.gradient_checkpointing_enable(gradient_checkpointing_kwargs={"use_reentrant": False})
```
Comment (abdulfatir) [inline multimodal/src/autogluon/multimodal/models/hf_text.py:133]: Was this by default `False` before?

File: core/src/autogluon/core/_setup_utils.py (hunk 1 above)
Comment (shchur) [reply core/src/autogluon/core/_setup_utils.py:34]: Updated the PR description with the pointers to the breaking changes

File: multimodal/src/autogluon/multimodal/models/hf_text.py (hunk 2 above)
Comment (shchur) [reply multimodal/src/autogluon/multimodal/models/hf_text.py:133]: It appears so https://github.com/huggingface/transformers/issues/26969

**Review by FANGAreNotGnu — APPROVED** [review]
LGTM to me, thanks! I'll add TODOs for the broken tutorials and we can decide if we would like to fix or deprecate before 1.5 release.
